File: ios/Normalize.swift
//
//  Normalize.swift
//  openlandTests
//
//  Created by Steve Kite on 5/9/19.
//  Copyright © 2019 Openland. All rights reserved.
//

import Foundation
import SwiftyJSON

fileprivate class SharedDictionary<K : Hashable, V> {
  var dict : Dictionary<K, V> = Dictionary()
  subscript(key : K) -> V? {
    get {
      return dict[key]
    }
    set(newValue) {
      dict[key] = newValue
    }
  }
}

fileprivate class NormalizedCollection {
  var records: [String: SharedDictionary<String, RecordValue>] = [:]
  func build() -> [String: SRecord] {
    var res: [String: SRecord] = [:]
    for k in records.keys {
      res[k] = SRecord(key: k, fields: records[k]!.dict)
    }
    return res
  }
}

fileprivate func normalizeValue(
  parentCacheKey: String?,
  collection: NormalizedCollection,
  value: OutputType,
  args: JSON,
  data: JSON
) -> RecordValue? {
  
  if (value is OutputType.NotNull) {
    let v2 = value as! OutputType.NotNull
    let res = normalizeValue(
      parentCacheKey: parentCacheKey,
      collection: collection,
      value: v2.inner,
      args: args,
      data: data
    )
    if res is RecordValue.NullValue {
      fatalError() // TODO: Replace
    } else {
      return res
    }
  }
  if (!data.exists() || data.null != nil) {
    return RecordValue.NullValue()
  }
  if value is OutputType.Scalar {
    let sc = value as! OutputType.Scalar
    if parentCacheKey != nil {
      if sc.name == "String" || sc.name == "Date" || sc.name == "ID" {
        return RecordValue.StringValue(value: data.stringValue)
      } else if sc.name == "Int" || sc.name == "Float" {
        if data.double != nil {
          return RecordValue.NumberValue(value: data.doubleValue)
        } else if data.int != nil {
          return RecordValue.NumberValue(value: Double(data.intValue))
        } else {
          fatalError() // TODO: Replace
        }
      } else if sc.name == "Boolean" {
        if data.bool != nil {
          return RecordValue.BooleanValue(value: data.boolValue)
        } else {
          fatalError() // TODO: Replace
        }
      }
    } else {
      return nil
    }
  } else if value is OutputType.List {
    let ls = value as! OutputType.List
    let vals = data.arrayValue
    var res: [RecordValue] = []
    for i in 0..<vals.count {
      res.append(normalizeValue(
        parentCacheKey: parentCacheKey! + "." + String(i),
        collection: collection,
        value: ls.inner,
        args: args,
        data:  vals[i]
      )!)
    }
    return RecordValue.ListValue(items: res)
  } else if value is OutputType.Object {
    let obj = value as! OutputType.Object
    return normalizeSelector(
      parentCacheKey: parentCacheKey,
      collection: collection,
      selectors: obj.selectors,
      args: args,
      data: data
    )
  }
  
  fatalError("Unreachable code")
}

fileprivate func normalizeSelector(
  parentCacheKey: String?,
  collection: NormalizedCollection,
  selectors: [Selector],
  args: JSON,
  data: JSON
) -> RecordValue? {
  
  var id: String? = nil
  if data["id"].exists() {
    if data["id"].null != nil /* is null */ {
      id = parentCacheKey
    } else {
      id = data["id"].stringValue
    }
  } else {
    id = parentCacheKey
  }
  var map: SharedDictionary<String, RecordValue>? = nil
  if id != nil {
    if collection.records[id!] != nil {
      map = collection.records[id!]!
    } else {
      map = SharedDictionary()
      collection.records[id!] = map!
    }
  }
  
  for f in selectors {
    if f is Selector.Field {
      let f2 = f as! Selector.Field
      if map != nil {
        let storeKey = f2.name
        map![storeKey] = normalizeValue(
          parentCacheKey: id!+"."+f2.name,
          collection: collection,
          value: f2.type,
          args: args,
          data: data[f2.alias]
        )
      } else {
        let _ = normalizeValue(
          parentCacheKey: nil,
          collection: collection,
          value: f2.type,
          args: args,
          data: data[f2.alias]
        )
      }
    } else if f is Selector.TypeCondition {
      let f2 = f as! Selector.TypeCondition
      if data["__typename"].string == f2.name {
        let _ = normalizeSelector(
          parentCacheKey: parentCacheKey,
          collection: collection,
          selectors: f2.type.selectors,
          args: args,
          data: data
        )
      }
    } else if f is Selector.Fragment {
      let f2 = f as! Selector.Fragment
      let _ = normalizeSelector(
        parentCacheKey: parentCacheKey,
        collection: collection,
        selectors: f2.type.selectors,
        args: args,
        data: data
      )
    }
  }
  
  if id != nil {
    return RecordValue.ReferenceValue(key: id!)
  } else {
    return nil
  }
}

func normalizeData(id: String, type: OutputType.Object, args: JSON, data: JSON) -> [String: SRecord] {
  let collection = NormalizedCollection()
  let _ = normalizeSelector(parentCacheKey: id, collection: collection, selectors: type.selectors, args: args, data: data)
  return collection.build()
}

File: packages/openland-graphql/spacex/store/normalize.ts
import { Selector, OutputType } from './../types';
import { OutputTypeObject } from '../types';
import { RecordValue, RecordSet, RecordValueReference } from './RecordStore';
import { selectorKey } from './selectorKey';

type RecordCollection = { [key: string]: { [key: string]: RecordValue } };

type NormalizeCtx = { collection: RecordCollection, queryArguments: any, root: any };

function buildSet(collection: RecordCollection): RecordSet {
    let keys = Object.keys(collection);
    let res: RecordSet = {};
    for (let k of keys) {
        res[k] = { key: k, fields: collection[k] };
    }
    return res;
}

function normalizeValue(parentCacheKey: string | null, value: OutputType, data: any, ctx: NormalizeCtx): RecordValue | null {
    if (value.type === 'notNull') {
        let res = normalizeValue(parentCacheKey, value.inner, data, ctx);
        if (res && res.type === 'null') {
            throw Error('Unexpected null value');
        } else {
            return res;
        }
    }
    if (data === null || data === undefined) {
        return { type: 'null' };
    }
    if (value.type === 'scalar') {
        if (parentCacheKey !== null) {
            if (value.name === 'String' || value.name === 'ID' || value.name === 'Date') {
                return { type: 'string', value: data + '' };
            } else if (value.name === 'Int' || value.name === 'Float') {
                if (typeof data === 'number') {
                    return { type: 'number', value: data };
                } else {
                    throw Error('Unexpected value for ' + value.name + ': ' + data);
                }
            } else if (value.name === 'Boolean') {
                if (typeof data === 'boolean') {
                    return { type: 'boolean', value: data };
                } else {
                    throw Error('Unexpected value for ' + value.name + ': ' + data);
                }
            } else {
                throw Error('Unsupported Scalar: ' + value.name);
            }
        } else {
            return null;
        }
    } else if (value.type === 'list') {
        if (!Array.isArray(data)) {
            throw Error('Invalid array');
        }
        if (parentCacheKey !== null) {
            let items: RecordValue[] = [];
            for (let i = 0; i < data.length; i++) {
                items.push(normalizeValue(`${parentCacheKey}.${i}`, value.inner, data[i], ctx)!);
            }
            return { type: 'list', values: items };
        } else {
            for (let i = 0; i < data.length; i++) {
                normalizeValue(null, value.inner, data[i], ctx);
            }
            return null;
        }
    } else if (value.type === 'object') {
        return normalizeSelector(parentCacheKey, value.selectors, data, ctx);
    }

    throw Error('Unreachable code');
}

function normalizeSelector(parentCacheKey: string | null, selectors: Selector[], data: any, ctx: NormalizeCtx): RecordValueReference | null {
    let map: { [key: string]: RecordValue } | undefined = undefined;
    let id: string | null = null;
    if (data.id !== null && data.id !== undefined) {
        let v = data.id;
        id = '' + v;
    } else {
        id = parentCacheKey;
    }
    if (id !== null) {
        let ex = ctx.collection[id];
        if (ex) {
            map = ex;
        } else {
            map = {};
            ctx.collection[id] = map;
        }
    }

    for (let f of selectors) {
        if (f.type === 'field') {
            if (map) {
                let key = selectorKey(f.name, f.arguments, ctx.queryArguments);
                map[key] = normalizeValue(id! + '.' + key, f.fieldType, data[f.alias], ctx)!;
            } else {
                normalizeValue(null, f.fieldType, data[f.alias], ctx);
            }
        } else if (f.type === 'type-condition') {
            if (data.__typename === f.name) {
                normalizeSelector(parentCacheKey, f.fragmentType.selectors, data, ctx);
            }
        } else if (f.type === 'fragment') {
            normalizeSelector(parentCacheKey, f.fragmentType.selectors, data, ctx);
        } else {
            throw Error('Unreachable code');
        }
    }

    if (id !== null) {
        return { type: 'reference', key: id };
    } else {
        return null;
    }
}

function normalizeRootSelector(rootCacheKey: string | null, selectors: Selector[], data: any, ctx: NormalizeCtx) {
    if (rootCacheKey !== null) {
        for (let f of selectors) {
            if (f.type !== 'field') {
                throw Error('Root query cant\'t contain fragments');
            }
            let key = selectorKey(f.name, f.arguments, ctx.queryArguments);
            let id = `${rootCacheKey}.${key}`;
            let refId = `${rootCacheKey}.\$ref.${key}`;
            let ex = ctx.collection[refId];
            let map: { [key: string]: RecordValue };
            if (!ex) {
                map = {};
                ctx.collection[refId] = map;
            } else {
                map = ex;
            }
            map.data = normalizeValue(id, f.fieldType, data[f.alias], ctx)!;
        }
    } else {
        normalizeSelector(null, selectors, data, ctx);
    }
}

export function normalizeData(id: string, type: OutputTypeObject, queryArguments: any, data: any): RecordSet {
    let collection: RecordCollection = {};
    normalizeSelector(id, type.selectors, data, { collection, root: data, queryArguments });
    return buildSet(collection);
}

export function normalizeResponse(rootCacheKey: string | null, type: OutputTypeObject, queryArguments: any, data: any): RecordSet {
    let collection: RecordCollection = {};
    normalizeRootSelector(rootCacheKey, type.selectors, data, { collection, root: data, queryArguments });
    return buildSet(collection);
}